code/game/objects/items/puzzle_pieces.dm

```diff
@@ -461,3 +522,99 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/puzzle_keycardpad, 32)
 		if(isnull(openclose))
 			openclose = door.density
 		INVOKE_ASYNC(door, openclose ? TYPE_PROC_REF(/obj/machinery/door/poddoor, open) : TYPE_PROC_REF(/obj/machinery/door/poddoor, close))
+
+#define MAX_PUZZLE_DOTS_PER_ROW 4
+#define PUZZLE_DOTS_VERTICAL_OFFSET 7
+#define PUZZLE_DOTS_HORIZONTAL_OFFSET 7
+
+///A dotted board that can be used as clue for PIN puzzle machinery
+/obj/effect/decal/puzzle_dots
+	name = "dotted board"
+	desc = "A board filled with colored dots. What could this mean?"
+	icon = 'icons/obj/fluff/puzzle_small.dmi'
+	icon_state = "puzzle_dots"
+	plane = GAME_PLANE //visible over walls
+	resistance_flags = INDESTRUCTIBLE | FIRE_PROOF | UNACIDABLE | LAVA_PROOF
+	flags_1 = UNPAINTABLE_1
+	///The id of the puzzle we're linked to.
+	var/id
+
+/obj/effect/decal/puzzle_dots/Initialize(mapload)
+	. = ..()
+	if(id)
+		SSqueuelinks.add_to_queue(src, id)
+
+/obj/effect/decal/puzzle_dots/MatchedLinks(id, partners)
+	var/obj/machinery/puzzle/password/pin/pad = locate() in partners
+	var/list/pass_digits = splittext(pad.password, "")
+	var/pass_len = length(pass_digits)
+	var/extra_rows = CEILING((pass_len/MAX_PUZZLE_DOTS_PER_ROW)-1, 1)
+	if(extra_rows)
+		pixel_y += round(extra_rows*(PUZZLE_DOTS_VERTICAL_OFFSET*0.5))
+		for(var/i in 1 to extra_rows)
+			var/mutable_appearance/row = mutable_appearance(icon, icon_state)
+			row.pixel_y = -i*PUZZLE_DOTS_VERTICAL_OFFSET
+			add_overlay(row)
+	for(var/i in 1 to pass_len)
+		var/mutable_appearance/colored_dot = mutable_appearance(icon, "puzzle_dot_single")
+		colored_dot.color = pad.digit_to_color[pass_digits[i]]
+		colored_dot.pixel_x = PUZZLE_DOTS_HORIZONTAL_OFFSET * ((i-1)%MAX_PUZZLE_DOTS_PER_ROW)
+		colored_dot.pixel_y -= CEILING((i/MAX_PUZZLE_DOTS_PER_ROW)-1, 1)*PUZZLE_DOTS_VERTICAL_OFFSET
+		add_overlay(colored_dot)
+
+#undef MAX_PUZZLE_DOTS_PER_ROW
+#undef PUZZLE_DOTS_VERTICAL_OFFSET
+#undef PUZZLE_DOTS_HORIZONTAL_OFFSET
+
+
+/obj/effect/decal/cleanable/crayon/puzzle
+	name = "graffiti that's clue to a puzzle"
+	icon_state = "0"
+	///The id of the puzzle we're linked to.
+	var/puzzle_id
+
+/obj/effect/decal/cleanable/crayon/puzzle/Initialize(mapload, main, type, e_name, graf_rot, alt_icon = null)
+	. = ..()
+	name = "number"
+	if(puzzle_id)
+		SSqueuelinks.add_to_queue(src, puzzle_id)
+
+/obj/effect/decal/cleanable/crayon/puzzle/MatchedLinks(id, partners)
+	var/obj/machinery/puzzle/password/pad = locate() in partners
+	var/list/pass_character = splittext(pad.password, "")
+	var/chosen_character = icon_state
+	if(!findtext(chosen_character, GLOB.is_alphanumeric))
+		qdel(src)
+		return FALSE
+	icon_state = pick(pass_character)
+	if(!text2num(icon_state))
+		name = "letter"
+		desc = "A letter vandalizing the station."
+	return TRUE
+
+/obj/effect/decal/cleanable/crayon/puzzle/pin
+	name = "graffiti that's clue to a PIN puzzle"
```

Comment:
this name feels really odd to read, can it be something like "a potential clue" or "PIN number"?